core/src/test/scala/integration/kafka/api/SaslPlainSslEndToEndAuthorizationTest.scala

```diff
@@ -16,41 +16,118 @@
 import org.junit.Test
 
 object SaslPlainSslEndToEndAuthorizationTest {
+
   class TestPrincipalBuilder extends KafkaPrincipalBuilder {
 
     override def build(context: AuthenticationContext): KafkaPrincipal = {
       context match {
         case ctx: SaslAuthenticationContext =>
           ctx.server.getAuthorizationID match {
-            case JaasTestUtils.KafkaPlainAdmin =>
+            case KafkaPlainAdmin =>
               new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "admin")
-            case JaasTestUtils.KafkaPlainUser =>
+            case KafkaPlainUser =>
               new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "user")
             case _ =>
               KafkaPrincipal.ANONYMOUS
           }
       }
     }
   }
+
+  object Credentials {
+    val allUsers = Map(KafkaPlainUser -> "user1-password",
+      KafkaPlainUser2 -> KafkaPlainPassword2,
+      KafkaPlainAdmin -> "broker-password")
+  }
+
+  class TestServerCallbackHandler extends AuthenticateCallbackHandler {
+    def configure(configs: java.util.Map[String, _], saslMechanism: String, jaasConfigEntries: java.util.List[AppConfigurationEntry]) {}
+    def handle(callbacks: Array[Callback]) {
+      var username: String = null
+      for (callback <- callbacks) {
+        if (callback.isInstanceOf[NameCallback])
+          username = callback.asInstanceOf[NameCallback].getDefaultName
+        else if (callback.isInstanceOf[PlainAuthenticateCallback]) {
+          val plainCallback = callback.asInstanceOf[PlainAuthenticateCallback]
+          plainCallback.authenticated(Credentials.allUsers(username) == new String(plainCallback.password))
+        } else
+          throw new UnsupportedCallbackException(callback)
+      }
+    }
+    def close() {}
+  }
+
+  class TestClientCallbackHandler extends AuthenticateCallbackHandler {
+    def configure(configs: java.util.Map[String, _], saslMechanism: String, jaasConfigEntries: java.util.List[AppConfigurationEntry]) {}
+    def handle(callbacks: Array[Callback]) {
+      val subject = Subject.getSubject(AccessController.getContext())
+      val username = subject.getPublicCredentials(classOf[String]).iterator().next()
+      for (callback <- callbacks) {
+        if (callback.isInstanceOf[NameCallback])
+          callback.asInstanceOf[NameCallback].setName(username)
+        else if (callback.isInstanceOf[PasswordCallback]) {
+          if (username == KafkaPlainUser || username == KafkaPlainAdmin)
+            callback.asInstanceOf[PasswordCallback].setPassword(Credentials.allUsers(username).toCharArray)
+        } else
+          throw new UnsupportedCallbackException(callback)
+      }
+    }
+    def close() {}
+  }
 }
 
+
+// This test uses SASL callback handler overrides for server connections of Kafka broker
+// and client connections of Kafka producers and consumers. Client connections of Kafka brokers
+// use default callback handlers. The second client used in the multi-user test
```

Comment:
"Client connections of Kafka brokers use default callback handlers." It seems that the clients are using customized callbacks?

Reply:
Broker use default client callback handlers for inter-broker communication. Producers/consumers are configured with customized callbacks. Updated the comment.